core/src/main/java/com/graphhopper/routing/util/spatialrules/SpatialRuleLookupJTS.java

```diff
@@ -0,0 +1,140 @@
+import org.locationtech.jts.geom.Polygon;
+import org.locationtech.jts.index.strtree.STRtree;
+
+import com.graphhopper.util.shapes.GHPoint;
+
+/**
+ * @author Thomas Butz
+ */
+public class SpatialRuleLookupJTS implements SpatialRuleLookup {
+    
+    private final GeometryFactory geometryFactory = new GeometryFactory();
+    private final List<SpatialRule> rules;
+    private final Envelope maxBounds;
+    private final STRtree index;
+    
+    public SpatialRuleLookupJTS(List<SpatialRule> spatialRules, Envelope maxBounds) {
+        this.index = new STRtree();
+        
+        Map<Polygon, SpatialRuleContainer> containerMap = new HashMap<>();
+        Set<String> ruleIDs = new HashSet<>();
+        for (SpatialRule rule : spatialRules) {
+            if (rule == null)
+                throw new IllegalArgumentException("rule cannot be null");
+
+            if (rule.equals(SpatialRule.EMPTY))
+                throw new IllegalArgumentException("rule cannot be EMPTY");
+            
+            if (!ruleIDs.add(rule.getId()))
+                throw new IllegalArgumentException("Duplicate rule ID: \"" + rule.getId() + "\"");
+            
+            for (Polygon border : rule.getBorders()) {
+                Envelope borderEnvelope = border.getEnvelopeInternal();
+                if (!maxBounds.intersects(borderEnvelope)) {
+                    continue;
+                }
+                
+                SpatialRuleContainer container = containerMap.get(border);
+                if (container == null) {
+                    container = new SpatialRuleContainer(border);
+                    containerMap.put(border, container);
+                    index.insert(borderEnvelope, container);
+                }
+                container.addRule(rule);
+            }
+        }
+
+        index.build();
+
+        this.rules = new ArrayList<>();
+        rules.add(SpatialRule.EMPTY);
+        rules.addAll(spatialRules);
+        this.maxBounds = maxBounds;
+    }
+
+    @Override
+    public SpatialRule lookupRule(double lat, double lon) {
+        if (!maxBounds.covers(lon, lat)) {
+            return SpatialRule.EMPTY;
+        }
+
+        Envelope searchEnv = new Envelope(lon, lon, lat, lat);
+        
+        @SuppressWarnings("unchecked")
+        List<SpatialRuleContainer> containers = index.query(searchEnv);
+        
+        if (containers.isEmpty()) {
+            return SpatialRule.EMPTY;
+        }
+        
+        Point point = geometryFactory.createPoint(new Coordinate(lon, lat));
+        List<SpatialRule> applicableRules = new ArrayList<>();
+        for (SpatialRuleContainer container : containers) {
+            if (container.containsProperly(point)) {
+                applicableRules.addAll(container.getRules());
+            }
+        }
+        
+        if (applicableRules.isEmpty()) {
+            return SpatialRule.EMPTY;
+        }
+
+        //TODO support multiple rules
+        return applicableRules.get(0);
```

Comment:
Since this would work out of the box now and since we completely change the implementation, wouldn't it make sense to break the interface? Or maybe deprecate the lookupRule method and create a new lookupRules method. lookupRule could fall back to returning only the first result.

Reply:
The interfaces still need a bit of love but that will be the focus of the following PRs. The next step is going to be adding support for multiple overlapping rules with priorities.

#1808 describes the roadmap

Reply:
Ok sure, that makes sense :)